File: backend/src/main/java/com/gpstore/catalog/shop/ShopCategory.java
package com.gpstore.catalog.shop;

import com.gpstore.platform.ShopOwned;
import com.gpstore.platform.ShopScopeFilter;
import com.gpstore.platform.TenantEntityListener;
import jakarta.persistence.*;
import org.hibernate.annotations.Filter;

import java.time.LocalDateTime;

/**
 * A department that belongs to ONE shop.
 *
 * <p>THE DISTINCTION THIS EXISTS TO KEEP. {@code categories} is the PLATFORM's
 * taxonomy - one tree every merchant picks from, where renaming a row renames
 * it for everybody. That is why writing to it needs CATALOG_DEFINE the moment a
 * second shop exists, and why a phone shop tapping Add Category was told "You
 * don't have permission to do that".
 *
 * <p>But the merchant was not trying to rewrite the marketplace. They were
 * trying to organise their own shelf - to have somewhere to put chargers. That
 * is a shopkeeper's daily work and it was impossible, because the only concept
 * in the schema was the shared one.
 *
 * <p>IT IS SHOP-OWNED, so everything the tenant layer already does applies with
 * no further work: reads are filtered to the shop in scope, a row belonging to
 * another shop is refused even when fetched by primary key, and an insert is
 * stamped with the shop the credential resolved to rather than any id the
 * object arrived carrying. Deepak Phone Shop creating "charger" therefore
 * cannot put "charger" on GUPT SAREE's screen, and cannot rename theirs.
 *
 * <p>{@link #globalCategoryId} IS OPTIONAL AND IS THE BRIDGE. Null means "my
 * own department, mine alone". Set means "this is me stocking that platform
 * department" - which is how enabling an existing category and inventing a new
 * one stay one concept instead of two that drift. It is also the controlled
 * route by which the platform can later promote a merchant's word into the
 * official tree: point the link at the new official row, keep this one.
 */
@Entity
@Table(name = "shop_categories")
@Filter(name = ShopScopeFilter.NAME, condition = ShopScopeFilter.CONDITION)
@EntityListeners(TenantEntityListener.class)
public class ShopCategory implements ShopOwned {

    @Id
    @GeneratedValue(strategy = GenerationType.IDENTITY)
    private Long id;

    @Column(name = "shop_id")
    private Long shopId;

    @Column(nullable = false, length = 120)
    private String name;

    @Column(length = 500)
    private String description;

    /** The platform department this one stands for, when it stands for one. */
    @Column(name = "global_category_id")
    private Long globalCategoryId;

    @Column(name = "display_order")
    private Integer displayOrder;

    @Column(nullable = false)
    private Boolean active = Boolean.TRUE;

    @Column(name = "created_at", nullable = false)
    private LocalDateTime createdAt;

    @Column(name = "updated_at", nullable = false)
    private LocalDateTime updatedAt;

    @PrePersist
    void onInsert() {
        LocalDateTime now = LocalDateTime.now();
        if (createdAt == null) {
            createdAt = now;
        }
        updatedAt = now;
        if (active == null) {
            active = Boolean.TRUE;
        }
    }

    @PreUpdate
    void onUpdate() {
        updatedAt = LocalDateTime.now();
    }

    @Override
    public Long getShopId() { return shopId; }

    @Override
    public void setShopId(Long shopId) { this.shopId = shopId; }

    public Long getId() { return id; }
    public void setId(Long id) { this.id = id; }

    public String getName() { return name; }
    public void setName(String name) { this.name = name; }

    public String getDescription() { return description; }
    public void setDescription(String description) { this.description = description; }

    public Long getGlobalCategoryId() { return globalCategoryId; }
    public void setGlobalCategoryId(Long globalCategoryId) { this.globalCategoryId = globalCategoryId; }

    public Integer getDisplayOrder() { return displayOrder; }
    public void setDisplayOrder(Integer displayOrder) { this.displayOrder = displayOrder; }

    public Boolean getActive() { return active; }
    public void setActive(Boolean active) { this.active = active; }

    public LocalDateTime getCreatedAt() { return createdAt; }
    public void setCreatedAt(LocalDateTime createdAt) { this.createdAt = createdAt; }

    public LocalDateTime getUpdatedAt() { return updatedAt; }
    public void setUpdatedAt(LocalDateTime updatedAt) { this.updatedAt = updatedAt; }
}

File: frontend/lib/features/admin/domain/shop_category.dart
import 'package:flutter/foundation.dart';

/// A department belonging to THIS shop.
///
/// NOT THE PLATFORM TAXONOMY. `Category` (in products/domain) is the shared
/// tree every merchant picks from, where renaming a row renames it for
/// everybody - which is why only the platform may write it, and why a phone
/// shop tapping Add Category was told "You don't have permission to do that".
///
/// This is the other thing the merchant actually wanted: somewhere of their
/// own to put chargers. Creating one adds a row this shop owns; no other
/// merchant sees it, and no customer's category tree grows an entry.
@immutable
class ShopCategory {
  const ShopCategory({
    required this.id,
    required this.name,
    this.description,
    this.globalCategoryId,
    this.displayOrder,
    this.active = true,
  });

  final int id;
  final String name;
  final String? description;

  /// The platform department this one stands for, when it stands for one.
  /// Null means "purely this shop's own".
  final int? globalCategoryId;

  final int? displayOrder;
  final bool active;

  factory ShopCategory.fromJson(Map<String, dynamic> json) => ShopCategory(
        id: (json['id'] as num).toInt(),
        name: (json['name'] ?? '').toString(),
        description: json['description'] as String?,
        globalCategoryId: (json['globalCategoryId'] as num?)?.toInt(),
        displayOrder: (json['displayOrder'] as num?)?.toInt(),
        active: json['active'] as bool? ?? true,
      );

  Map<String, dynamic> toJson() => {
        'id': id,
        'name': name,
        'description': description,
        'globalCategoryId': globalCategoryId,
        'displayOrder': displayOrder,
        'active': active,
      };

  @override
  bool operator ==(Object other) => other is ShopCategory && other.id == id;

  @override
  int get hashCode => id.hashCode;
}
